HeterogeneousCore/AlpakaCore/interface/MoveToDeviceCache.h

```diff
@@ -0,0 +1,101 @@
+#ifndef HeterogeneousCore_AlpakaInterface_interface_MoveToDeviceCache_h
+#define HeterogeneousCore_AlpakaInterface_interface_MoveToDeviceCache_h
+
+#include <type_traits>
+
+#include <alpaka/alpaka.hpp>
+
+#include "HeterogeneousCore/AlpakaCore/interface/QueueCache.h"
+#include "HeterogeneousCore/AlpakaCore/interface/CopyToDeviceCache.h"
+#include "HeterogeneousCore/AlpakaInterface/interface/devices.h"
+
+namespace cms::alpakatools {
+  namespace detail {
+    // By default copy the host object with CopyToDevice<T>
+    //
+    // Doing with template specialization (rather than
+    // std::conditional_t and if constexpr) because the
+    // CopyToDevice<THostObject>::copyAsync() is ill-defined e.g. for
+    // PortableCollection on host device
+    template <typename TDevice, typename THostObject>
+    class MoveToDeviceCacheImpl {
+    public:
+      using HostObject = THostObject;
+      using Impl = CopyToDeviceCacheImpl<TDevice, THostObject>;
+      using DeviceObject = typename Impl::DeviceObject;
+
+      MoveToDeviceCacheImpl(HostObject&& srcObject) : impl_(srcObject) {}
+
+      DeviceObject const& get(size_t i) const { return impl_.get(i); }
+
+    private:
+      Impl impl_;
+    };
+
+    // For host device, move the host object instead
+    template <typename THostObject>
+    class MoveToDeviceCacheImpl<alpaka_common::DevHost, THostObject> {
+    public:
+      using HostObject = THostObject;
+      using DeviceObject = HostObject;
+
+      MoveToDeviceCacheImpl(HostObject&& srcObject) : data_(std::move(srcObject)) {}
+
+      DeviceObject const& get(size_t i) const { return data_; }
+
+    private:
+      HostObject data_;
+    };
+  }  // namespace detail
+
+  /**
+   * This class template implements a cache for data that is moved
+   * from the host (of type THostObject) to all the devices
+   * corresponding to the TDevice device type.
+   *
+   * The host-side object to be moved is given as an argument to the
+   * class constructor. The constructor uses the
+   * CopyToDevice<THostObject> class template to copy the data to the
+   * devices, and waits for the data copies to finish, i.e. the
+   * constructor is synchronous wrt. the data copies. The "move" is
+   * achieved by requiring the constructor argument to the rvalue
```

Comment:
```suggestion
   * achieved by requiring the constructor argument to be an rvalue
```
?

Reply:
By the way, after re-reading the description and the implementation, this works because
  - for non-CPU backends we actually make N copies
  - for the CPU backend we can move, because there is exactly one device

So, if alpaka ever implements N devices for the CPU back-ends (_e.g._ one per socket or one per NUMA node) we would need to revisit the logic, right ?

Reply:
> ?

Thanks for catching.

Reply:
> By the way, after re-reading the description and the implementation, this works because
> 
> * for non-CPU backends we actually make N copies
> * for the CPU backend we can move, because there is exactly one device

Correct.

> So, if alpaka ever implements N devices for the CPU back-ends (_e.g._ one per socket or one per NUMA node) we would need to revisit the logic, right ?

In presence of multiple CPU device objects the present code results in one object, and the `MoveToDeviceCache::get()` returns a `const` reference to that cached object for all CPU device objects. This happens because the `DevHost` specialization of the implementation ignores the device native handle (index).

So unless we'd _want_ to have a cached object for each CPU device object, I believe the present code should be able to handle that situation.

Reply:
Ah, yes, you're right of course.